boot-speed/clouds/measure-cloud.py

```diff
@@ -0,0 +1,242 @@
+    res = stream.query(filters)
+    id = res[0]['id']
+    return id
+
+
+def measure_instance(instance, datadir, reboots=1):
+    print("Measuring instance", instance.id)
+
+    # Use the same command (and hence format) used when measuring devices
+    os.system("date --utc --rfc-3339=ns > " +
+              str(Path(datadir, "job-start-timestamp")))
+
+    # Do not refresh the snaps for the moment.
+    # Regular Ubuntu Server images do not auto-reboot on snap refreshes as Core
+    # does, but we want to keep the measurement scripts as similar as possible.
+
+    instance.execute(
+        'sudo snap set system refresh.hold='
+        '"$(date --date=tomorrow +%Y-%m-%dT%H:%M:%S%:z)"')
+
+    instance.execute(
+        "wget https://raw.githubusercontent.com/CanonicalLtd/"
+        "server-test-scripts/master/boot-speed/bootspeed.sh")
+    instance.execute("chmod +x bootspeed.sh")
+    bootid = instance.execute("cat /proc/sys/kernel/random/boot_id")
+    print("boot_id:", bootid)
+
+    instance.execute("rm -rf artifacts")
+    outstr = instance.execute("./bootspeed.sh 2>&1")
+    print(outstr)
+    outstr = instance.execute("find artifacts")
+    print(outstr)
+
+    # Test for the existence of the file bootspeed.sh creates if it reached to
+    # the end of the measurement with no errors.
+    outstr = instance.execute("test -f artifacts/measurement-successful"
+                              "&& echo ok")
+    if outstr != "ok":
+        print("Measurement failed (missing measurement-successful)!")
+        sys.exit(1)
+
+    instance.execute("mv artifacts boot_0")
+    instance.execute("tar czf boot_0.tar.gz boot_0")
+
+    sftpclient = create_sftp_client(
+        instance.ip, 'ubuntu', keyfilepath=instance.key_pair.private_key_path)
+    sftpclient.get("boot_0.tar.gz", "boot_0.tar.gz")
+    sftpclient.close()
+
+    instance.execute("sudo snap refresh")
+
+    for nboot in range(1, reboots+1):
+        bootdir = "boot_" + str(nboot)
+        instance.restart()
+        new_bootid = instance.execute("cat /proc/sys/kernel/random/boot_id")
+        if new_bootid == bootid:
+            print("Cloud instance did not reboot!")
+            sys.exit(1)
+        bootid = new_bootid
+        instance.execute("rm -rf artifacts")
+        outstr = instance.execute("./bootspeed.sh 2>&1")
+        print(outstr)
+        instance.execute("mv artifacts " + bootdir)
+        instance.execute("tar czf " + bootdir + ".tar.gz " + bootdir)
+
+        sftpclient = create_sftp_client(
+            instance.ip, 'ubuntu',
+            keyfilepath=instance.key_pair.private_key_path)
+        sftpclient.get(bootdir + ".tar.gz", bootdir + ".tar.gz")
+        sftpclient.close()
+
+    for tarball in glob.glob('boot_*.tar.gz'):
+        with tarfile.open(tarball, "r:gz") as tar:
+            tar.extractall(path=str(datadir))
+        os.unlink(tarball)
+
+
+def measure_ec2(
+        release, datadir, *, inst_type="t2.micro", region="us-east-1",
```

Comment:
\* for?

Reply:
To force the parameters following the * (`inst_type` and `region` in this case) to be keyword-only, and not positional. While developing it happened to me to use them in the wrong order, so I decided to force the usage the `param=value` syntax when calling the function. That's all; the * is not necessary, it just makes the syntax a bit more strict.